src/hdmf/validate/validator.py

```diff
@@ -503,75 +532,170 @@ def __construct_illegal_link_error(self, child_spec, parent_builder):
     def __cannot_be_link(spec):
         return not isinstance(spec, LinkSpec) and not spec.linkable
 
-    def __get_child_validator(self, spec):
-        """Returns the appropriate validator for a child spec
+    def __get_child_validator(self, builder, spec):
+        """Returns the appropriate validator for a child builder
 
-        If a specific data type can be resolved, the validator is acquired from
-        the ValidatorMap, otherwise a new Validator is created.
+        If a specific data type of the builder can be resolved, the validator
+        is acquired from the ValidatorMap, otherwise a new Validator is created
+        from the spec.
         """
-        if _resolve_data_type(spec) is not None:
-            return self.vmap.get_validator(_resolve_data_type(spec))
+        base_spec = _resolve_spec_links(spec, self.vmap)
+        dt = _resolve_builder_data_type(builder, base_spec)
+        if dt is not None:
+            return self.vmap.get_validator(dt)
         elif isinstance(spec, GroupSpec):
             return GroupValidator(spec, self.vmap)
         elif isinstance(spec, DatasetSpec):
             return DatasetValidator(spec, self.vmap)
         else:
-            msg = "Unable to resolve a validator for spec %s" % spec
+            msg = "Unable to resolve a validator for builder %s" % builder
             raise ValueError(msg)
 
+    def __warn_on_extra_fields(self, extra_builders):
+        """Warn for each child builder which is not part of the spec
+
+        Extra fields are any children attached to the parent builder
+        which are not part of the parent spec.
+        """
+        for builder in extra_builders:
+            name_of_erroneous = self.get_spec_loc(self.spec)
+            builder_loc = self.get_builder_loc(builder)
+            yield ExtraFieldWarning(name_of_erroneous, location=builder_loc)
 
-class SpecMatches:
-    """A utility class to hold a spec and the builders matched to it"""
 
-    def __init__(self, spec):
+class SpecMatch:
+    """A utility class class to hold a spec and the fields matched to it"""
```

Comment:
I suggest adding methods for manipulating this object. Currently, it looks like it's left up to the user to know how to properly manipulate the underlying contents of the object.

Reply:
Good suggestion. You're right, usage of `SpecMatch` isn't clear. Hopefully this new implementation is an improvement!